File: src/com/suushiemaniac/cubing/bld/util/MapUtil.java
package com.suushiemaniac.cubing.bld.util;

import java.util.*;
import java.util.function.Function;
import java.util.function.Supplier;
import java.util.stream.Collectors;

public class MapUtil {
	public static <K extends Comparable<? super K>, V> void sortedMapPrint(Map<K, V> toPrint) {
		List<K> sortedKeys = new ArrayList<>(toPrint.keySet());
		sortedKeys.sort(Comparator.naturalOrder());

		for (K i : sortedKeys) {
			System.out.println(i.toString() + ": " + toPrint.get(i));
		}
	}

	public static <T extends Number> float freqMapAverage(Map<Integer, T> toAvg) {
		float sum = 0;
		int criteriaHit = 0;

		for (int i : toAvg.keySet()) {
			Number mapEntry = toAvg.get(i);
			criteriaHit += mapEntry.intValue();
			sum += i * mapEntry.floatValue();
		}

		return sum / criteriaHit;
	}

	public static <T, V> Map<T, V> constantValueMap(Collection<T> keys, Supplier<V> value) {
		return keys
				.stream()
				.collect(Collectors.toMap(Function.identity(), k -> value.get(), (a, b) -> b));
	}
}

File: src/com/suushiemaniac/cubing/bld/util/MapUtil.kt
package com.suushiemaniac.cubing.bld.util

object MapUtil {
    fun <K : Comparable<K>, V> Map<K, V>.sortedPrint() {
        for (i in this.keys.sorted()) {
            println("$i: ${this[i]}")
        }
    }

    fun <T : Number> Map<T, Int>.freqAverage(): Double {
        val criteriaHit = this.values.sum()
        val sum = this.entries.sumByDouble { it.key.toDouble() * it.value }

        return sum / criteriaHit
    }

    infix fun <K, V> Collection<K>.allTo(value: (K) -> V): Map<K, V> {
        return this.map { it to value(it) }.toMap()
    }

    infix fun <K, V> Collection<K>.alwaysTo(value: () -> V): Map<K, V> {
        return this.map { it to value() }.toMap()
    }

    infix fun <K, V> Collection<K>.alwaysTo(value: V): Map<K, V> {
        return this.map { it to value }.toMap()
    }

    fun <K, V> MutableMap<K, V>.reset(action: (Map.Entry<K, V>) -> V) {
        for (entry in this) {
            this[entry.key] = action(entry)
        }
    }

    fun <K> MutableMap<K, Int>.increment(key: K) {
        this[key] = this.getOrDefault(key, 0) + 1
    }

    fun <K, V> Map<K, V?>.denullify(): Map<K, V> {
        return this.mapNotNull { (key, nullableVal) ->
            nullableVal?.let { key to it }
        }.toMap()
    }
}
